maas/resource_maas_boot_resource.go

```diff
@@ -0,0 +1,330 @@
+package maas
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/canonical/gomaasclient/client"
+	"github.com/canonical/gomaasclient/entity"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func ResourceMaasBootResources() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Provides a resource to manage MAAS bootresources.",
+		CreateContext: resourceBootResourcesCreate,
+		ReadContext:   resourceBootResourcesRead,
+		UpdateContext: resourceBootResourcesUpdate,
+		DeleteContext: resourceBootResourcesDelete,
+
+		Schema: map[string]*schema.Schema{
+			"boot_source": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "The boot source database ID this resource set is associated with.",
+			},
+			"boot_source_selections": {
+				Type:        schema.TypeSet,
+				Required:    true,
+				Description: "The set of database IDs for boot source selections to attach to this boot resource",
+				Elem:        &schema.Schema{Type: schema.TypeInt},
+			},
+		},
+	}
+}
+
+func resourceBootResourcesCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	err := awaitImportComplete(client)
+	if err != nil {
+		return diag.Errorf("Could not await image importing: %v", err)
+	}
+
+	resources, err := getBootResources(client, "synced")
+	if err != nil {
+		return diag.Errorf("error fetching synced boot resources: %v", err)
+	}
+	resourceMap := make(map[string]struct{})
+	for _, res := range resources {
+		resourceMap[res.Name] = struct{}{}
+	}
+
+	bootselections := d.Get("boot_source_selections").(*schema.Set).List()
+
+	if len(bootselections) == 0 {
+		return diag.Errorf("At least one boot source selection must be added to the boot resources")
+	}
+	bootsource, err := getBootSource(client)
+	if err != nil {
+		return diag.Errorf("error fetching boot source: %v", err)
+	}
+	d.Set("boot_source", bootsource.ID)
+	d.SetId(fmt.Sprintf("%v", bootsource.ID))
+
+	var selectionSet []int
+	for _, bootselection := range bootselections {
+		bootselection, err := getBootSourceSelection(client, bootsource.ID, bootselection.(int))
+		if err != nil {
+			return diag.Errorf("error fetching boot selection %v: %s", bootselection, err)
+		}
+
+		// check the selection has it's resource created
+		if _, exists := resourceMap[fmt.Sprintf("%s/%s", bootselection.OS, bootselection.Release)]; !exists {
+			return diag.Errorf("Boot Resource missing for %s/%s", bootselection.OS, bootselection.Release)
+		}
+		selectionSet = append(selectionSet, bootselection.ID)
+	}
+	if err := d.Set("boot_source_selections", unique(selectionSet)); err != nil {
```

Comment:
if unique is ensured by validating user's input, then this unique converter is not needed.